File: playwright-python/pages/cart_page.py
from decimal import Decimal

from playwright.sync_api import Page, expect


class CartPage:
    def __init__(self, page: Page, base_url: str):
        self.page = page
        self.base_url = base_url.rstrip("/")
        self.cart = page.locator("#cart")
        self.last_cart_response_status: int | None = None

    def open(self) -> None:
        self.page.goto(self.base_url)
        expect(self.page.locator("#product-grid")).to_have_attribute(
            "aria-busy", "false"
        )

    def product_card(self, product_name: str):
        return self.page.locator(".product-card").filter(has_text=product_name)

    def cart_line(self, product_name: str):
        return self.page.locator("#cart-items .cart-item").filter(
            has_text=product_name
        )

    def add_product(self, product_name: str, quantity: int = 1) -> None:
        card = self.product_card(product_name)
        if card.count() == 0:
            self.open()
            card = self.product_card(product_name)
        expect(card).to_have_count(1)
        quantity_input = card.locator(".quantity-input")
        quantity_input.fill(str(quantity))
        quantity_input.dispatch_event("change")
        card.get_by_role("button", name="Add to cart", exact=True).click()
        expect(self.page.locator("#app-toast")).to_contain_text(
            f"Added {quantity}"
        )
        expect(self.page.locator("#app-toast")).to_contain_text(product_name)
        expect(self.cart_line(product_name)).to_be_visible()

    def select_quantity(self, product_name: str, quantity: int) -> None:
        card = self.product_card(product_name)
        quantity_input = card.locator(".quantity-input")
        quantity_input.fill(str(quantity))
        quantity_input.dispatch_event("change")
        expect(quantity_input).to_have_value(str(quantity))

    def add_selected_product(self, product_name: str) -> None:
        card = self.product_card(product_name)
        quantity = int(card.locator(".quantity-input").input_value())
        card.get_by_role("button", name="Add to cart", exact=True).click()
        expect(self.page.locator("#app-toast")).to_contain_text(
            f"Added {quantity}"
        )
        expect(self.page.locator("#app-toast")).to_contain_text(product_name)
        expect(self.cart_line(product_name)).to_be_visible()

    def expect_cart_hidden(self) -> None:
        expect(self.cart).to_be_hidden()

    def expect_cart_visible(self) -> None:
        expect(self.cart).to_be_visible()

    def expect_product_in_cart(self, product_name: str) -> None:
        expect(self.cart_line(product_name)).to_be_visible()

    def expect_product_not_in_cart(self, product_name: str) -> None:
        expect(self.cart_line(product_name)).to_have_count(0)

    def expect_total_items(self, quantity: int) -> None:
        expect(self.page.locator("#cart-total-items")).to_have_text(str(quantity))

    def expect_grand_total(self, total: Decimal) -> None:
        expect(self.page.locator("#cart-grand-total")).to_have_text(
            f"${total:.2f}"
        )

    def expect_line_quantity(self, product_name: str, quantity: int) -> None:
        expect(self.cart_line(product_name).locator(".muted")).to_contain_text(
            f"{quantity} "
        )

    def expect_one_line(self, product_name: str) -> None:
        expect(self.cart_line(product_name)).to_have_count(1)

    def scroll_to_top(self) -> None:
        self.page.evaluate("window.scrollTo(0, document.body.scrollHeight)")

    def expect_cart_inside_viewport(self) -> None:
        box = self.cart.bounding_box()
        viewport = self.page.viewport_size
        assert box is not None, "Floating cart has no visible bounding box"
        assert viewport is not None, "Browser viewport size is unavailable"
        assert box["x"] >= 0 and box["y"] >= 0
        assert box["x"] + box["width"] <= viewport["width"]
        assert box["y"] + box["height"] <= viewport["height"]

    def remove_product(self, product_name: str) -> None:
        line = self.cart_line(product_name)
        line.get_by_role("button", name="Remove", exact=True).click()
        expect(line).to_have_count(0)

    def clear(self) -> None:
        self.page.get_by_role("button", name="Clear", exact=True).click()
        self.expect_cart_hidden()

    def reload(self) -> None:
        self.page.reload()
        expect(self.page.locator("#product-grid")).to_have_attribute(
            "aria-busy", "false"
        )

    def attempt_to_add_over_stock(self, product_name: str, stock: int) -> None:
        card = self.product_card(product_name)
        quantity_input = card.locator(".quantity-input")
        attempted_quantity = stock + 1
        quantity_input.evaluate(
            "(element, value) => { element.max = value; element.value = value; }",
            str(attempted_quantity),
        )
        with self.page.expect_response(
            lambda response: response.url.rstrip("/").endswith("/cart")
            and response.request.method == "POST"
        ) as response_info:
            card.get_by_role("button", name="Add to cart", exact=True).click()
        self.last_cart_response_status = response_info.value.status

    def expect_request_rejected(self) -> None:
        assert self.last_cart_response_status == 400

    def expect_stock_error(self, available: int, product_name: str) -> None:
        expect(self.page.locator("#cart-message")).to_have_text(
            f"Only {available} units left of {product_name}"
        )

    def expect_catalog_stock(self, product_name: str, stock: int) -> None:
        expect(self.product_card(product_name).locator(".badge")).to_have_text(
            f"{stock} in stock"
        )

File: playwright-typescript/src/pages/cart-page.ts
import assert from "node:assert/strict";
import { expect, type Locator, type Page } from "@playwright/test";
import { settings } from "../support/settings.js";

export class CartPage {
  private readonly cart: Locator;
  lastCartResponseStatus?: number;

  constructor(private readonly page: Page) {
    this.cart = page.locator("#cart");
  }

  get cartItems(): Locator {
    return this.page.locator("#cart-items .cart-item");
  }

  productCard(name: string): Locator {
    return this.page.locator(".product-card").filter({ hasText: name });
  }

  cartLine(name: string): Locator {
    return this.cartItems.filter({ hasText: name });
  }

  async open(): Promise<void> {
    await this.page.goto(settings.baseUrl);
    await expect(this.page.locator("#product-grid")).toHaveAttribute("aria-busy", "false");
  }

  async addProduct(name: string, quantity = 1): Promise<void> {
    let card = this.productCard(name);
    if (await card.count() === 0) {
      await this.open();
      card = this.productCard(name);
    }
    await expect(card).toHaveCount(1);
    const input = card.locator(".quantity-input");
    await input.fill(String(quantity));
    await input.dispatchEvent("change");
    await card.getByRole("button", { name: "Add to cart", exact: true }).click();
    await expect(this.page.locator("#app-toast")).toContainText(`Added ${quantity}`);
    await expect(this.page.locator("#app-toast")).toContainText(name);
    await expect(this.cartLine(name)).toBeVisible();
  }

  async selectQuantity(name: string, quantity: number): Promise<void> {
    const input = this.productCard(name).locator(".quantity-input");
    await input.fill(String(quantity));
    await input.dispatchEvent("change");
    await expect(input).toHaveValue(String(quantity));
  }

  async addSelectedProduct(name: string): Promise<void> {
    const card = this.productCard(name);
    const quantity = Number(await card.locator(".quantity-input").inputValue());
    await card.getByRole("button", { name: "Add to cart", exact: true }).click();
    await expect(this.page.locator("#app-toast")).toContainText(`Added ${quantity}`);
    await expect(this.cartLine(name)).toBeVisible();
  }

  expectHidden(): Promise<void> { return expect(this.cart).toBeHidden(); }
  expectVisible(): Promise<void> { return expect(this.cart).toBeVisible(); }
  expectProductInCart(name: string): Promise<void> { return expect(this.cartLine(name)).toBeVisible(); }
  expectProductNotInCart(name: string): Promise<void> { return expect(this.cartLine(name)).toHaveCount(0); }
  expectTotalItems(quantity: number): Promise<void> { return expect(this.page.locator("#cart-total-items")).toHaveText(String(quantity)); }
  expectGrandTotal(total: number): Promise<void> { return expect(this.page.locator("#cart-grand-total")).toHaveText(`$${total.toFixed(2)}`); }
  expectLineQuantity(name: string, quantity: number): Promise<void> { return expect(this.cartLine(name).locator(".muted")).toContainText(`${quantity} `); }
  expectOneLine(name: string): Promise<void> { return expect(this.cartLine(name)).toHaveCount(1); }
  scroll(): Promise<unknown> { return this.page.evaluate(() => window.scrollTo(0, document.body.scrollHeight)); }

  async expectInsideViewport(): Promise<void> {
    const box = await this.cart.boundingBox();
    const viewport = this.page.viewportSize();
    assert.ok(box && viewport);
    assert.ok(box.x >= 0 && box.y >= 0);
    assert.ok(box.x + box.width <= viewport.width);
    assert.ok(box.y + box.height <= viewport.height);
  }

  async removeProduct(name: string): Promise<void> {
    const line = this.cartLine(name);
    await line.getByRole("button", { name: "Remove", exact: true }).click();
    await expect(line).toHaveCount(0);
  }

  async clear(): Promise<void> {
    await this.page.getByRole("button", { name: "Clear", exact: true }).click();
    await this.expectHidden();
  }

  async reload(): Promise<void> {
    await this.page.reload();
    await expect(this.page.locator("#product-grid")).toHaveAttribute("aria-busy", "false");
  }

  async attemptOverStock(name: string, stock: number): Promise<void> {
    const card = this.productCard(name);
    const attempted = String(stock + 1);
    await card.locator(".quantity-input").evaluate((element: HTMLInputElement, value) => {
      element.max = String(value);
      element.value = String(value);
    }, attempted);
    const responsePromise = this.page.waitForResponse((response) => response.url().replace(/\/$/, "").endsWith("/cart") && response.request().method() === "POST");
    await card.getByRole("button", { name: "Add to cart", exact: true }).click();
    this.lastCartResponseStatus = (await responsePromise).status();
  }

  expectRequestRejected(): void { assert.equal(this.lastCartResponseStatus, 400); }
  expectStockError(stock: number, name: string): Promise<void> { return expect(this.page.locator("#cart-message")).toHaveText(`Only ${stock} units left of ${name}`); }
  expectCatalogStock(name: string, stock: number): Promise<void> { return expect(this.productCard(name).locator(".badge")).toHaveText(`${stock} in stock`); }
}
